_posts/2023-10-23-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,314 @@
+The UTO improves on scalability and maintenance, but also provides support for Kafka clusters running in KRaft mode.
+It is fully backwards compatible with the BTO, but it will only reconcile topic events in one direction, from Kubernetes to Kafka.
+Your `KafkaTopic` custom resources will be the source-of-truth, like with all the other operators.
+
+The operator detects cases where multiple `KafkaTopic` resources are attempting to manage a Kafka topic using the same `.spec.topicName`.
+Only the oldest resource will be reconciled, while the other resource will fail with a `ResourceConflict` error.
+
+We still do not support decreasing `.spec.partitions` and changing `.spec.replicas`.
+While the former is not supported by Kafka itself, support for the latter could be added in the future.
+
+In addition to be able to pause reconciliations, the operator now also provides a way to unmanage a Kafka topic using annotations.
+If the `KafkaTopic` is paused and we delete it, then it will be also deleted in Kafka.
+If the `KafkaTopic` is unmanaged and we delete it, then only the resource will be garbage collected by Kubernetes, but the topic will still exists in Kafka.
+
+In terms of requirements, the UTO assumes the following access rights:
+
+- describe all topics
+- describe all topics configs
+- create topics
+- create partitions
+- delete topics
+- list partition reassignments
+- describe broker configs
+
+#### Scalability
+
+The BTO is limited in terms of scalability, as it only operates on one topic at a time, and it has to update a persistent store containing topic metadata.
+Instead, the UTO does not store topic metadata and it aims to be scalable in terms of the number of topics that it can operate on.
+The following line graph confirms the BTO scalability issue, while the UTO scales almost linearly.
+
+<figure>
+    <img src="/assets/images/posts/2023-10-23-uto-max-recon-time.png" height=350>
+    <figcaption><small>
+        Fig 2. Line graph comparing BTO and UTO max reconciliation times.<br/>
+        We used a 3-nodes cluster with default configurations, running on a local Minikube instance with 10 cores and 28 GB of memory.<br/>
+        The test driver application was running on the same machine, and the time spent on the event queue was not included.
+    </small></figcaption>
+</figure>
+
+When running Kafka operations, the UTO makes use of the request batching supported by the Kafka Admin client to get higher throughput for metadata operations.
+All `KafkaTopic` events are queued when received, and then processed in batches by a number of controller threads (currently, only a single thread is supported).
+If more than one event related to a single topic resource are found in the batch building, they are put back in the queue to be processed with the next batch.
+
+You can tune the batching mechanism by setting `STRIMZI_MAX_QUEUE_SIZE` (default: 1024), `STRIMZI_MAX_BATCH_SIZE` (default: 100), and `MAX_BATCH_LINGER_MS` (default: 100).
```

Comment:
Just an idea, which comes to my mind. I think it would be great to try to do some experiments with these parameters + (deploy 3 different clusters) and see what the results are and talk about it in the next blog post. WDYT?

Reply:
Yes. I already did some experiments using a custom Java load test driver, but don't want to add too much to this presentation. A follow up blog post about tuning would be great, also considering that we are adding some [additional metrics](https://github.com/strimzi/strimzi-kafka-operator/pull/9216) that can be used for that.